src/Microsoft.Diagnostics.Monitoring.RestServer/ProducesWithProblemDetailsAttribute.cs

```diff
@@ -0,0 +1,46 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.AspNetCore.Mvc.Filters;
+using System;
+
+namespace Microsoft.Diagnostics.Monitoring.RestServer
+{
+    /// <summary>
+    /// Overrides the behavior of the Produces attribute, which clears all content types from the result and
+    /// sets the content types from the attribute to the result. This attribute does the same except that if
+    /// the result is a Bad Request, then only add the application/problem+json content type to the result.
+    /// </summary>
+    /// <remarks>
+    /// This alleviates an issue where application/problem+json would never be returned if some other content
+    /// type, such as application/json, was also specified. The DefaultOutputFormatterSelector will format the
+    /// content using any acceptable content type, which both application/json and application/problem+json are
+    /// acceptable to the SystemTextJsonOutputFormatter, thus it chooses the first one: application/json. This
+    /// selection is incorrect for error results, which typically want application/problem+json.
+    /// 
+    /// The intent of this attribute is to make sure that the output is unconditionally reported with a content
+    /// type of application/problem+json if the result is a Bad Request.
+    /// </remarks>
+    [AttributeUsage(AttributeTargets.Class | AttributeTargets.Method, AllowMultiple = false, Inherited = true)]
+    public class ProducesWithProblemDetailsAttribute : ProducesAttribute
+    {
+        public ProducesWithProblemDetailsAttribute(string contentType, params string[] additionalContentTypes)
+            : base(contentType, additionalContentTypes)
+        {
+        }
+
+        public override void OnResultExecuting(ResultExecutingContext context)
+        {
+            if (context.Result is BadRequestObjectResult badRequestResult && badRequestResult.Value is ProblemDetails)
+            {
+                badRequestResult.ContentTypes.Add(RestServer.ContentTypes.ApplicationProblemJson);
```

Comment:
Does `[Result].ContentTypes` describe all the MIME types of the current response or all possible responses from this API.

Reply:
It describes the allowable content types of the returned response instance. There are other attributes that can modify the applicable list of content types (typically its just the ProducesAttribute, but we're overriding the behavior here).